content/posts/kong-dynamic-plugin-ordering.adoc

```diff
@@ -0,0 +1,170 @@
+
+In Kong Gateway 3.0 we're introducing a _dynamic plugin ordering_ functionality.
+You can specify a directed graph of plugins.
+With this, we should be able to specify config that `rate-limiting` should run before `key-auth` or `oath2-introspection`.
+
+== Example: Default behavior before Kong 3.0
+
+.I have Kong Gateway 3.0 Enterprise instance running locally.
+image::kong_gw_3_0_insomnia.png[]
+
+.A service points to a random Wow expression by Owen Wilson.
+image::create_wow_service_insomnia.png[]
+
+.A route `/wow` points to `wow-service`
+image::create_route_insomnia.png[]
+
+[source,bash]
+.filename
+----
+$ curl -v localhost:8000  #<1>
+
+$ curl --request POST \
+  --url http://localhost:8001/services/ \
+  --data name=wow-service \
+  --data url=https://owen-wilson-wow-api.herokuapp.com/wows/random  #<2>
+
+$ curl --request POST \
+  --url http://localhost:8001/services/wow-service/routes \
+  --data paths=/wow   #<3>
+
+$ curl -i -X POST http://localhost:8001/plugins \
+    --data name=key-auth  #<4>
+
+$ curl -i -X POST \
+    --url http://localhost:8001/consumers/ \
+    --data "username=example-user"  #<5>
+
+$ curl -i -X POST \
+  --url http://localhost:8001/consumers/example-user/key-auth/ \
+  --data 'key=test'   #<6>
+
+$ curl -i -X POST http://localhost:8001/plugins \
+    --data name=rate-limiting \
+    --data config.minute=1 \
+    --data config.policy=local  #<7>
+----
+<1> Kong Gateway 3.0 up and running
+<2> A service points to a random Wow expression by Owen Wilson.
+<3> A route `/wow` points to `wow-service`.
+<4> Create a `key-auth` plugin instance.
+<5> Create `example-user` consumer.
+<6> Create apiKey `test`
+<7> Create a global rate limiting plugin.
+
+Here is a default flow: API user calls `:8000/wow` to access API.
+Kong (via `key-auth` plugin) is there is an API key in the header.
+If there is one, it validates it and passes next to the `rate-limiting` plugin.
+If there is no apiKey in the header, Kong throws an error.
+
+[source,json]
+.error.json
+----
+{
+    "message": "No API key found in request"
+}
+----
+
+== A solution implemented in Kong Gateway 3.0
+
+Next, let's modify this behavior to invoke `rate-limiting` regardless of whether a key is provided or not.
+
+[source,bash]
+.filename.sh
+----
+RLID=`curl localhost:8001/plugins | jq -r '.data[] | select(.name == "rate-limiting").id'`
+
+curl -i -X DELETE http://localhost:8001/plugins/$RLID
+
+curl -i -X POST http://localhost:8001/plugins \
+  --data name=rate-limiting \
+  --data config.minute=1 \
+  --data config.policy=local \
+  --data ordering.before.access=key-auth    #<1>
+----
+<1> an essential new config `ordering.before.access` defines what phase of the specified plugin this plugin needs to be invoked.
+
+Now, when we invoke the `:8000/wow` API, we will get `No API key found in request,` but if we introspect the headers, we will see that the `rate-limiting` plugin is up for the task already.
```

Comment:
"inspect" instead of "introspect"

Reply:
Showing these headers is great. Maybe in the steps above, before you enabled ordering, you could show the result of the same request _without_ the RL headers to drive the point home clearly